File: load_testing/locust/mixed_workload.py
"""
Mixed Workload Load Test for PsychSync
Simulates realistic traffic patterns with multiple user types
"""

from locust import HttpUser, task, between, events
import logging
import random
from datetime import datetime

from locust_config import (
    LoadTestConfig,
    get_headers,
    test_data_manager,
)

logger = logging.getLogger(__name__)


class PsychSyncUser(HttpUser):
    """
    Simulates realistic PsychSync usage patterns:
    - 15% Authentication tasks
    - 40% Assessment taking
    - 20% Dashboard viewing
    - 15% Team management
    - 5% Assessment management
    - 5% AI/NLP features
    """

    wait_time = between(1, 4)

    def on_start(self):
        """Setup: Login and initialize user state"""
        self.token = None
        self.user_credentials = test_data_manager.get_random_user()
        self.current_assessment_id = None
        self.team_id = test_data_manager.get_random_team_id()
        self.login()

    def login(self):
        """Login to get access token"""
        credentials = self.user_credentials

        with self.client.post(
            "/api/v1/auth/token-fixed",
            data={
                "username": credentials["email"],
                "password": credentials["password"],
            },
            headers=get_headers(),
            catch_response=True,
            name="Mixed: Login",
        ) as response:
            if response.status_code == 200:
                data = response.json()
                self.token = data.get("access_token")
                response.success()
            else:
                response.failure(f"Login failed: {response.status_code}")

    # ==================== AUTHENTICATION TASKS (15% weight) ====================

    @task(3)
    def refresh_token(self):
        """Task: Refresh authentication token"""
        with self.client.post(
            "/api/v1/auth/refresh",
            json={"refresh_token": "dummy_refresh_token"},
            headers=get_headers(),
            catch_response=True,
            name="Auth: Token Refresh",
        ) as response:
            if response.status_code in [200, 401]:
                response.success()
            else:
                response.failure(f"Token refresh failed: {response.status_code}")

    @task(2)
    def verify_user(self):
        """Task: Verify current user"""
        with self.client.get(
            "/api/v1/users/me",
            headers=get_headers(self.token),
            catch_response=True,
            name="Auth: Verify User",
        ) as response:
            if response.status_code == 200:
                response.success()
            else:
                response.failure(f"Verify user failed: {response.status_code}")

    # ==================== ASSESSMENT TASKS (40% weight) ====================

    @task(10)
    def browse_assessments(self):
        """Task: Browse available assessment frameworks"""
        with self.client.get(
            "/api/v1/personality-assessments/frameworks",
            headers=get_headers(self.token),
            catch_response=True,
            name="Assessment: Browse Frameworks",
        ) as response:
            if response.status_code == 200:
                response.success()
            else:
                response.failure(f"Browse failed: {response.status_code}")

    @task(8)
    def start_assessment(self):
        """Task: Start a new assessment"""
        framework = test_data_manager.get_random_framework()
        assessment_id = test_data_manager.get_random_assessment_id()
        self.current_assessment_id = assessment_id

        with self.client.post(
            f"/api/v1/assessments/{assessment_id}/start",
            headers=get_headers(self.token),
            json={"framework": framework},
            catch_response=True,
            name="Assessment: Start",
        ) as response:
            if response.status_code in [200, 201]:
                response.success()
            else:
                response.failure(f"Start failed: {response.status_code}")

    @task(15)
    def submit_assessment_responses(self):
        """Task: Submit assessment responses (auto-save)"""
        if not self.current_assessment_id:
            self.current_assessment_id = test_data_manager.get_random_assessment_id()

        responses = [
            {
                "question_id": f"q{random.randint(1, 100)}",
                "answer": random.randint(1, 5),
            }
            for _ in range(random.randint(1, 5))
        ]

        with self.client.post(
            f"/api/v1/assessments/{self.current_assessment_id}/responses",
            headers=get_headers(self.token),
            json={"responses": responses},
            catch_response=True,
            name="Assessment: Submit Responses",
        ) as response:
            if response.status_code in [200, 201]:
                response.success()
            else:
                response.failure(f"Submit failed: {response.status_code}")

    @task(5)
    def view_assessment_results(self):
        """Task: View assessment results"""
        assessment_id = self.current_assessment_id or test_data_manager.get_random_assessment_id()

        with self.client.get(
            f"/api/v1/assessments/{assessment_id}/results",
            headers=get_headers(self.token),
            catch_response=True,
            name="Assessment: View Results",
        ) as response:
            if response.status_code in [200, 404]:
                response.success()
            else:
                response.failure(f"View results failed: {response.status_code}")

    # ==================== DASHBOARD TASKS (20% weight) ====================

    @task(5)
    def load_dashboard(self):
        """Task: Load main dashboard"""
        with self.client.get(
            "/api/v1/analytics/dashboard",
            headers=get_headers(self.token),
            catch_response=True,
            name="Dashboard: Load Overview",
        ) as response:
            if response.status_code == 200:
                response.success()
            else:
                response.failure(f"Dashboard load failed: {response.status_code}")

    @task(4)
    def load_team_analytics(self):
        """Task: Load team analytics"""
        team_id = self.team_id

        with self.client.get(
            f"/api/v1/analytics/team/{team_id}",
            headers=get_headers(self.token),
            catch_response=True,
            name="Dashboard: Team Analytics",
        ) as response:
            if response.status_code == 200:
                response.success()
            else:
                response.failure(f"Team analytics failed: {response.status_code}")

    @task(3)
    def load_analytics_trends(self):
        """Task: Load historical trends"""
        with self.client.get(
            "/api/v1/analytics/trends",
            headers=get_headers(self.token),
            params={"days": random.choice([7, 30, 90])},
            catch_response=True,
            name="Dashboard: Trends",
        ) as response:
            if response.status_code == 200:
                response.success()
            else:
                response.failure(f"Trends failed: {response.status_code}")

    @task(2)
    def export_report(self):
        """Task: Export analytics report"""
        with self.client.get(
            "/api/v1/analytics/export",
            headers=get_headers(self.token),
            params={
                "format": random.choice(["pdf", "csv", "json"]),
                "team_id": self.team_id,
            },
            catch_response=True,
            name="Dashboard: Export Report",
        ) as response:
            if response.status_code in [200, 202]:
                response.success()
            else:
                response.failure(f"Export failed: {response.status_code}")

    # ==================== TEAM MANAGEMENT TASKS (15% weight) ====================

    @task(4)
    def view_team(self):
        """Task: View team details"""
        team_id = self.team_id

        with self.client.get(
            f"/api/v1/teams/{team_id}",
            headers=get_headers(self.token),
            catch_response=True,
            name="Team: View Details",
        ) as response:
            if response.status_code == 200:
                response.success()
            else:
                response.failure(f"View team failed: {response.status_code}")

    @task(3)
    def view_team_members(self):
        """Task: View team members"""
        team_id = self.team_id

        with self.client.get(
            f"/api/v1/teams/{team_id}/members",
            headers=get_headers(self.token),
            catch_response=True,
            name="Team: View Members",
        ) as response:
            if response.status_code == 200:
                response.success()
            else:
                response.failure(f"View members failed: {response.status_code}")

    @task(2)
    def view_team_activity(self):
        """Task: View team activity log"""
        team_id = self.team_id

        with self.client.get(
            f"/api/v1/teams/{team_id}/activity",
            headers=get_headers(self.token),
            catch_response=True,
            name="Team: View Activity",
        ) as response:
            if response.status_code == 200:
                response.success()
            else:
                response.failure(f"View activity failed: {response.status_code}")

    @task(1)
    def update_team_permissions(self):
        """Task: Update team permissions"""
        team_id = self.team_id

        with self.client.put(
            f"/api/v1/teams/{team_id}/permissions",
            headers=get_headers(self.token),
            json={
                "permissions": {
                    "view_analytics": random.choice([True, False]),
                    "manage_assessments": random.choice([True, False]),
                }
            },
            catch_response=True,
            name="Team: Update Permissions",
        ) as response:
            if response.status_code in [200, 403]:
                response.success()
            else:
                response.failure(f"Update permissions failed: {response.status_code}")

    # ==================== ASSESSMENT MANAGEMENT TASKS (5% weight) ====================

    @task(2)
    def list_assessments(self):
        """Task: List available assessments"""
        with self.client.get(
            "/api/v1/assessments",
            headers=get_headers(self.token),
            catch_response=True,
            name="Assessment Mgmt: List",
        ) as response:
            if response.status_code == 200:
                response.success()
            else:
                response.failure(f"List assessments failed: {response.status_code}")

    @task(1)
    def duplicate_assessment(self):
        """Task: Duplicate an assessment template"""
        assessment_id = test_data_manager.get_random_assessment_id()

        with self.client.post(
            f"/api/v1/assessments/{assessment_id}/duplicate",
            headers=get_headers(self.token),
            json={"name": f"Copy of {assessment_id}"},
            catch_response=True,
            name="Assessment Mgmt: Duplicate",
        ) as response:
            if response.status_code in [200, 201, 403]:
                response.success()
            else:
                response.failure(f"Duplicate failed: {response.status_code}")

    # ==================== AI/NLP TASKS (5% weight) ====================

    @task(2)
    def analyze_text(self):
        """Task: Submit text for NLP analysis"""
        sample_texts = [
            "I feel confident in my ability to lead teams effectively.",
            "Communication is key to successful project outcomes.",
            "I prefer working independently rather than in groups.",
            "I enjoy solving complex problems and analytical challenges.",
        ]

        with self.client.post(
            "/api/v1/nlp/analyze",
            headers=get_headers(self.token),
            json={
                "text": random.choice(sample_texts),
                "analysis_type": random.choice(["sentiment", "personality", "behavioral"]),
            },
            catch_response=True,
            name="AI: Text Analysis",
        ) as response:
            if response.status_code in [200, 202]:
                response.success()
            else:
                response.failure(f"Analysis failed: {response.status_code}")

    @task(1)
    def get_ai_insights(self):
        """Task: Get AI-generated insights"""
        assessment_id = test_data_manager.get_random_assessment_id()

        with self.client.get(
            f"/api/v1/ai/insights/{assessment_id}",
            headers=get_headers(self.token),
            catch_response=True,
            name="AI: Get Insights",
        ) as response:
            if response.status_code in [200, 404]:
                response.success()
            else:
                response.failure(f"Get insights failed: {response.status_code}")


# ==================== PERFORMANCE MONITORING ====================

@events.request.add_hook
def on_request(request_type, name, response_time, response_length, exception, **kwargs):
    """Monitor performance and log anomalies"""
    if exception:
        logger.error(f"Request failed: {name} - {exception}")
    elif response_time > LoadTestConfig.THRESHOLDS["p99"]:
        logger.warning(
            f"Slow request detected: {name} - {response_time}ms "
            f"(p99 threshold: {LoadTestConfig.THRESHOLDS['p99']}ms)"
        )


@events.spawning_complete.add_hook
def on_spawning_complete(user_count, **kwargs):
    """Log when all users have been spawned"""
    logger.info(f"Spawning complete: {user_count} users are now running")


@events.test_stop.add_hook
def on_test_stop(environment, **kwargs):
    """Log comprehensive test summary"""
    stats = environment.stats

    logger.info("\n" + "=" * 80)
    logger.info("MIXED WORKLOAD LOAD TEST SUMMARY")
    logger.info("=" * 80)
    logger.info(f"Total Requests: {stats.total.num_requests:,}")
    logger.info(f"Failures: {stats.total.num_failures:,}")
    logger.info(f"Failure Rate: {(stats.total.num_failures / stats.total.num_requests * 100):.2f}%")
    logger.info(f"Median Response Time: {stats.total.median_response_time}ms")
    logger.info(f"Average Response Time: {stats.total.avg_response_time:.1f}ms")
    logger.info(f"Min Response Time: {stats.total.min_response_time}ms")
    logger.info(f"Max Response Time: {stats.total.max_response_time}ms")
    logger.info(f"Requests/Second: {stats.total.total_rps:.2f}")
    logger.info("=" * 80)

    # Log top 5 slowest endpoints
    logger.info("\nTop 5 Slowest Endpoints:")
    sorted_stats = sorted(
        [s for s in stats.entries.values() if s.num_requests > 0],
        key=lambda x: x.avg_response_time,
        reverse=True,
    )[:5]

    for entry in sorted_stats:
        logger.info(
            f"  {entry.name}: {entry.avg_response_time:.1f}ms "
            f"({entry.num_requests} requests)"
        )

    logger.info("=" * 80 + "\n")


if __name__ == "__main__":
    import sys

    users = int(sys.argv[1]) if len(sys.argv) > 1 else 1000
    print(f"Starting mixed workload test with {users} users...")
    print("Target API:", LoadTestConfig.API_BASE_URL)
    print("\nTask Distribution:")
    print("  - Authentication: 15%")
    print("  - Assessment Taking: 40%")
    print("  - Dashboard & Analytics: 20%")
    print("  - Team Management: 15%")
    print("  - Assessment Management: 5%")
    print("  - AI/NLP Processing: 5%")
    print()

File: tests/performance/k6/mixed_workload.js
/**
 * Mixed Workload Load Test for PsychSync using k6
 * Simulates realistic traffic patterns with multiple user types
 */

import http from 'k6/http';
import { check, group, sleep } from 'k6';
import { Rate, Trend, Counter } from 'k6/metrics';

// Custom metrics
const errorRate = new Rate('errors');
const requestRate = new Rate('request_success');
const responseTime = new Trend('response_time');

// Endpoints metrics
const authMetrics = {
  success: new Rate('auth_success'),
  latency: new Trend('auth_latency'),
};

const assessmentMetrics = {
  success: new Rate('assessment_success'),
  latency: new Trend('assessment_latency'),
};

const dashboardMetrics = {
  success: new Rate('dashboard_success'),
  latency: new Trend('dashboard_latency'),
};

const teamMetrics = {
  success: new Rate('team_success'),
  latency: new Trend('team_latency'),
};

export const options = {
  scenarios: {
    mixed_workload: {
      executor: 'ramping-vus',
      startVUs: 0,
      stages: [
        { duration: '2m', target: 100 },   // Ramp up to 100 users
        { duration: '5m', target: 100 },   // Stay at 100 users
        { duration: '3m', target: 1000 },  // Ramp up to 1000 users
        { duration: '10m', target: 1000 }, // Stay at 1000 users
        { duration: '5m', target: 5000 },  // Ramp up to 5000 users
        { duration: '10m', target: 5000 }, // Stay at 5000 users
        { duration: '3m', target: 0 },     // Ramp down
      ],
      gracefulRampDown: '30s',
      tags: { scenario: 'mixed_workload' },
    },
  },
  thresholds: {
    'errors': ['rate<0.01'],
    'http_req_duration': ['p(95)<1000', 'p(99)<2000'],
    'auth_success': ['rate>0.95'],
    'assessment_success': ['rate>0.98'],
    'dashboard_success': ['rate>0.95'],
  },
};

const BASE_URL = __ENV.API_BASE_URL || 'http://localhost:8000';
const API_PREFIX = '/api/v1';

// Test data
const ASSESSMENT_IDS = ['mbti-001', 'big-five-001', 'enneagram-001', 'disc-001'];
const TEAM_IDS = ['team-001', 'team-002', 'team-003'];
const FRAMEWORKS = ['mbti', 'big_five', 'enneagram', 'predictive_index', 'disct'];

const TEST_USERS = Array.from({ length: 10000 }, (_, i) => ({
  email: `loadtest_user_${i}@test.com`,
  password: 'LoadTest123!',
}));

function getRandomUser() {
  return TEST_USERS[Math.floor(Math.random() * TEST_USERS.length)];
}

function getRandomItem(arr) {
  return arr[Math.floor(Math.random() * arr.length)];
}

// Authentication flow
function authenticateFlow() {
  const user = getRandomUser();
  const headers = {
    'Content-Type': 'application/json',
    'User-Agent': 'PsychSync-LoadTest-k6/1.0',
  };

  group('Authentication', () => {
    // Login
    const loginRes = http.post(
      `${BASE_URL}${API_PREFIX}/auth/token-fixed`,
      `username=${encodeURIComponent(user.email)}&password=${encodeURIComponent(user.password)}`,
      {
        headers: { ...headers, 'Content-Type': 'application/x-www-form-urlencoded' },
        tags: { name: 'Auth:Login' },
      }
    );

    const loginSuccess = check(loginRes, {
      'login status is 200': (r) => r.status === 200,
      'login has token': (r) => r.json('access_token') !== undefined,
    });

    authMetrics.success.add(loginSuccess);
    authMetrics.latency.add(loginRes.timings.duration);
    errorRate.add(!loginSuccess);

    if (loginSuccess) {
      const token = loginRes.json('access_token');
      const authHeaders = { ...headers, 'Authorization': `Bearer ${token}` };

      // Token refresh (15% weight)
      if (Math.random() < 0.15) {
        const refreshRes = http.post(
          `${BASE_URL}${API_PREFIX}/auth/refresh`,
          JSON.stringify({ refresh_token: 'dummy' }),
          {
            headers: authHeaders,
            tags: { name: 'Auth:Refresh' },
          }
        );

        check(refreshRes, {
          'refresh successful': (r) => [200, 401].includes(r.status),
        });

        errorRate.add(refreshRes.status !== 200 && refreshRes.status !== 401);
      }

      return authHeaders;
    }
  });

  return null;
}

// Assessment flow (40% weight)
function assessmentFlow(authHeaders) {
  if (!authHeaders) return;

  group('Assessment Taking', () => {
    // Browse frameworks (10%)
    if (Math.random() < 0.10) {
      const browseRes = http.get(
        `${BASE_URL}${API_PREFIX}/personality-assessments/frameworks`,
        { headers: authHeaders, tags: { name: 'Assessment:Browse' } }
      );

      check(browseRes, { 'browse successful': (r) => r.status === 200 });
      assessmentMetrics.success.add(browseRes.status === 200);
      assessmentMetrics.latency.add(browseRes.timings.duration);
      errorRate.add(browseRes.status !== 200);
    }

    // Start assessment (8%)
    if (Math.random() < 0.08) {
      const framework = getRandomItem(FRAMEWORKS);
      const assessmentId = getRandomItem(ASSESSMENT_IDS);

      const startRes = http.post(
        `${BASE_URL}${API_PREFIX}/assessments/${assessmentId}/start`,
        JSON.stringify({ framework }),
        { headers: authHeaders, tags: { name: 'Assessment:Start' } }
      );

      check(startRes, { 'start successful': (r) => [200, 201].includes(r.status) });
      assessmentMetrics.success.add([200, 201].includes(startRes.status));
      assessmentMetrics.latency.add(startRes.timings.duration);
      errorRate.add(!check(startRes, { 'start ok': (r) => [200, 201].includes(r.status) }));
    }

    // Submit responses (15% - most frequent)
    if (Math.random() < 0.15) {
      const assessmentId = getRandomItem(ASSESSMENT_IDS);
      const responses = Array.from({ length: 5 }, (_, i) => ({
        question_id: `q${Math.floor(Math.random() * 100)}`,
        answer: Math.floor(Math.random() * 5) + 1,
      }));

      const submitRes = http.post(
        `${BASE_URL}${API_PREFIX}/assessments/${assessmentId}/responses`,
        JSON.stringify({ responses }),
        { headers: authHeaders, tags: { name: 'Assessment:Submit' } }
      );

      check(submitRes, { 'submit successful': (r) => [200, 201].includes(r.status) });
      assessmentMetrics.success.add([200, 201].includes(submitRes.status));
      assessmentMetrics.latency.add(submitRes.timings.duration);
      errorRate.add(!check(submitRes, { 'submit ok': (r) => [200, 201].includes(r.status) }));
    }

    // View results (7%)
    if (Math.random() < 0.07) {
      const assessmentId = getRandomItem(ASSESSMENT_IDS);

      const resultsRes = http.get(
        `${BASE_URL}${API_PREFIX}/assessments/${assessmentId}/results`,
        { headers: authHeaders, tags: { name: 'Assessment:Results' } }
      );

      check(resultsRes, { 'results successful': (r) => [200, 404].includes(r.status) });
      assessmentMetrics.success.add([200, 404].includes(resultsRes.status));
      assessmentMetrics.latency.add(resultsRes.timings.duration);
      errorRate.add(!check(resultsRes, { 'results ok': (r) => [200, 404].includes(r.status) }));
    }

    sleep(Math.random() * 2 + 1);
  });
}

// Dashboard flow (20% weight)
function dashboardFlow(authHeaders) {
  if (!authHeaders || Math.random() > 0.20) return;

  group('Dashboard & Analytics', () => {
    // Load dashboard (8%)
    if (Math.random() < 0.08) {
      const dashboardRes = http.get(
        `${BASE_URL}${API_PREFIX}/analytics/dashboard`,
        { headers: authHeaders, tags: { name: 'Dashboard:Load' } }
      );

      check(dashboardRes, { 'dashboard successful': (r) => r.status === 200 });
      dashboardMetrics.success.add(dashboardRes.status === 200);
      dashboardMetrics.latency.add(dashboardRes.timings.duration);
      errorRate.add(dashboardRes.status !== 200);
    }

    // Team analytics (6%)
    if (Math.random() < 0.06) {
      const teamId = getRandomItem(TEAM_IDS);

      const teamRes = http.get(
        `${BASE_URL}${API_PREFIX}/analytics/team/${teamId}`,
        { headers: authHeaders, tags: { name: 'Dashboard:Team' } }
      );

      check(teamRes, { 'team analytics successful': (r) => r.status === 200 });
      dashboardMetrics.success.add(teamRes.status === 200);
      dashboardMetrics.latency.add(teamRes.timings.duration);
      errorRate.add(teamRes.status !== 200);
    }

    // Trends (4%)
    if (Math.random() < 0.04) {
      const days = [7, 30, 90][Math.floor(Math.random() * 3)];

      const trendsRes = http.get(
        `${BASE_URL}${API_PREFIX}/analytics/trends?days=${days}`,
        { headers: authHeaders, tags: { name: 'Dashboard:Trends' } }
      );

      check(trendsRes, { 'trends successful': (r) => r.status === 200 });
      dashboardMetrics.success.add(trendsRes.status === 200);
      dashboardMetrics.latency.add(trendsRes.timings.duration);
      errorRate.add(trendsRes.status !== 200);
    }

    // Export report (2%)
    if (Math.random() < 0.02) {
      const format = ['pdf', 'csv', 'json'][Math.floor(Math.random() * 3)];

      const exportRes = http.get(
        `${BASE_URL}${API_PREFIX}/analytics/export?format=${format}`,
        { headers: authHeaders, tags: { name: 'Dashboard:Export' } }
      );

      check(exportRes, { 'export successful': (r) => [200, 202].includes(r.status) });
      dashboardMetrics.success.add([200, 202].includes(exportRes.status));
      dashboardMetrics.latency.add(exportRes.timings.duration);
      errorRate.add(!check(exportRes, { 'export ok': (r) => [200, 202].includes(r.status) }));
    }

    sleep(Math.random() * 2 + 1);
  });
}

// Team management flow (15% weight)
function teamFlow(authHeaders) {
  if (!authHeaders || Math.random() > 0.15) return;

  group('Team Management', () => {
    const teamId = getRandomItem(TEAM_IDS);

    // View team (6%)
    if (Math.random() < 0.06) {
      const viewRes = http.get(
        `${BASE_URL}${API_PREFIX}/teams/${teamId}`,
        { headers: authHeaders, tags: { name: 'Team:View' } }
      );

      check(viewRes, { 'view team successful': (r) => r.status === 200 });
      teamMetrics.success.add(viewRes.status === 200);
      teamMetrics.latency.add(viewRes.timings.duration);
      errorRate.add(viewRes.status !== 200);
    }

    // View members (4%)
    if (Math.random() < 0.04) {
      const membersRes = http.get(
        `${BASE_URL}${API_PREFIX}/teams/${teamId}/members`,
        { headers: authHeaders, tags: { name: 'Team:Members' } }
      );

      check(membersRes, { 'members successful': (r) => r.status === 200 });
      teamMetrics.success.add(membersRes.status === 200);
      teamMetrics.latency.add(membersRes.timings.duration);
      errorRate.add(membersRes.status !== 200);
    }

    // View activity (3%)
    if (Math.random() < 0.03) {
      const activityRes = http.get(
        `${BASE_URL}${API_PREFIX}/teams/${teamId}/activity`,
        { headers: authHeaders, tags: { name: 'Team:Activity' } }
      );

      check(activityRes, { 'activity successful': (r) => r.status === 200 });
      teamMetrics.success.add(activityRes.status === 200);
      teamMetrics.latency.add(activityRes.timings.duration);
      errorRate.add(activityRes.status !== 200);
    }

    // Update permissions (2%)
    if (Math.random() < 0.02) {
      const updateRes = http.put(
        `${BASE_URL}${API_PREFIX}/teams/${teamId}/permissions`,
        JSON.stringify({
          permissions: {
            view_analytics: Math.random() > 0.5,
            manage_assessments: Math.random() > 0.5,
          },
        }),
        { headers: authHeaders, tags: { name: 'Team:Update' } }
      );

      check(updateRes, { 'update successful': (r) => [200, 403].includes(r.status) });
      teamMetrics.success.add([200, 403].includes(updateRes.status));
      teamMetrics.latency.add(updateRes.timings.duration);
      errorRate.add(!check(updateRes, { 'update ok': (r) => [200, 403].includes(r.status) }));
    }

    sleep(Math.random() * 2 + 1);
  });
}

// AI/NLP flow (5% weight)
function aiFlow(authHeaders) {
  if (!authHeaders || Math.random() > 0.05) return;

  group('AI/NLP Processing', () => {
    // Text analysis (3%)
    if (Math.random() < 0.03) {
      const texts = [
        'I feel confident in my ability to lead teams effectively.',
        'Communication is key to successful project outcomes.',
        'I prefer working independently rather than in groups.',
      ];

      const analyzeRes = http.post(
        `${BASE_URL}${API_PREFIX}/nlp/analyze`,
        JSON.stringify({
          text: texts[Math.floor(Math.random() * texts.length)],
          analysis_type: ['sentiment', 'personality'][Math.floor(Math.random() * 2)],
        }),
        { headers: authHeaders, tags: { name: 'AI:Analyze' } }
      );

      check(analyzeRes, { 'analysis successful': (r) => [200, 202].includes(r.status) });
      errorRate.add(!check(analyzeRes, { 'analysis ok': (r) => [200, 202].includes(r.status) }));
    }

    // Get insights (2%)
    if (Math.random() < 0.02) {
      const assessmentId = getRandomItem(ASSESSMENT_IDS);

      const insightsRes = http.get(
        `${BASE_URL}${API_PREFIX}/ai/insights/${assessmentId}`,
        { headers: authHeaders, tags: { name: 'AI:Insights' } }
      );

      check(insightsRes, { 'insights successful': (r) => [200, 404].includes(r.status) });
      errorRate.add(!check(insightsRes, { 'insights ok': (r) => [200, 404].includes(r.status) }));
    }

    sleep(Math.random() * 3 + 2);
  });
}

// Main test
export default function() {
  const authHeaders = authenticateFlow();

  // Execute different flows based on weights
  // 40% Assessment, 20% Dashboard, 15% Team, 5% AI
  const rand = Math.random();

  if (rand < 0.40) {
    assessmentFlow(authHeaders);
  } else if (rand < 0.60) {
    dashboardFlow(authHeaders);
  } else if (rand < 0.75) {
    teamFlow(authHeaders);
  } else {
    aiFlow(authHeaders);
  }

  // Base think time
  sleep(Math.random() * 2 + 1);
}

export function setup() {
  console.log(`Starting mixed workload test against: ${BASE_URL}`);
  console.log('Task Distribution:');
  console.log('  - Authentication: 15%');
  console.log('  - Assessment Taking: 40%');
  console.log('  - Dashboard & Analytics: 20%');
  console.log('  - Team Management: 15%');
  console.log('  - AI/NLP Processing: 10%');

  return { startTime: new Date().toISOString() };
}

export function teardown(data) {
  console.log('\n=== TEST SUMMARY ===');
  console.log(`Start: ${data.startTime}`);
  console.log(`End: ${new Date().toISOString()}`);
  console.log('====================\n');
}
